File: Medium/2026-01-07-1339-Maximum-Product-of-Splitted-Binary-Tree/solution.java
class Solution {
    private long totalSum = 0;
    private long maxProduct = 0;
    private final int MOD = 1000000007;

    public int maxProduct(TreeNode root) {
        totalSum = calculateSum(root);
        calculateMaxProduct(root);
        return (int) (maxProduct % MOD);
    }

    private long calculateSum(TreeNode node) {
        if (node == null) return 0;
        long leftSum = calculateSum(node.left);
        long rightSum = calculateSum(node.right);
        return node.val + leftSum + rightSum;
    }

    private long calculateMaxProduct(TreeNode node) {
        if (node == null) return 0;
        long leftSum = calculateMaxProduct(node.left);
        long rightSum = calculateMaxProduct(node.right);
        long currentSum = node.val + leftSum + rightSum;

        if (leftSum > 0) {
            maxProduct = Math.max(maxProduct, leftSum * (totalSum - leftSum));
        }
        if (rightSum > 0) {
            maxProduct = Math.max(maxProduct, rightSum * (totalSum - rightSum));
        }
        return currentSum;
    }
}

File: Medium/2026-01-07-1339-Maximum-Product-of-Splitted-Binary-Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val === undefined ? 0 : val);
 *     this.left = (left === undefined ? null : left);
 *     this.right = (right === undefined ? null : right);
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxProduct = function(root) {
    const MOD = 1e9 + 7;
    let totalSum = 0;
    let maxProd = 0;

    const calculateSum = (node) => {
        if (!node) return 0;
        const leftSum = calculateSum(node.left);
        const rightSum = calculateSum(node.right);
        const currentSum = leftSum + rightSum + node.val;
        totalSum += currentSum;
        return currentSum;
    };

    const findMaxProduct = (node) => {
        if (!node) return 0;
        const leftSum = findMaxProduct(node.left);
        const rightSum = findMaxProduct(node.right);
        const currentSum = leftSum + rightSum + node.val;
        const otherSum = totalSum - currentSum;
        maxProd = Math.max(maxProd, currentSum * otherSum);
        return currentSum;
    };

    calculateSum(root);
    findMaxProduct(root);
    
    return maxProd % MOD;
};